File: wolframsynth.py
from contextlib import closing
import wolframalpha
import boto3
import sys
import os

wolfram_app_id = "APPID-WOLFRAM"
wolfram_client = wolframalpha.Client(wolfram_app_id)
query = str(sys.argv[1])

print(":: Question: " + query + "\n")
print("++ Getting Answer from WolframAlpha...\n")
wolfram_response = wolfram_client.query(query)
wolfram_answer = next(wolfram_response.results).text.encode('ascii','ignore')

print(":: Answer:")
print(wolfram_answer)

cwd = os.getcwd()
aws_client = boto3.client('polly')
print("\n++ Getting synthesized audio file from Amazon Polly...\n")
aws_response = aws_client.synthesize_speech(
  OutputFormat='pcm',
  SampleRate='16000',
  Text="for the question " + query + ", the answer is: " + wolfram_answer + ".",
  TextType='text',
  VoiceId='Joanna',
 )

if "AudioStream" in aws_response:
    with closing(aws_response["AudioStream"]) as stream:
        output = cwd + "/answer.pcm"

        try:
            with open(output, "wb") as file:
                file.write(stream.read())
        except IOError as error:
            print(error)
            sys.exit(-1)

print("++ Playing the synthesized audio file..\n")
os.system("play -t raw -r 16k -e signed -b 16 -c 1 " + cwd + "/answer.pcm")

File: wolframsynth.rb
require 'aws-sdk'
require 'wolfram-alpha'

wolfram_options = { "format" => "plaintext" }
wolfram_client = WolframAlpha::Client.new "APPID-WOLFRAM", wolfram_options
wolfram_question = ARGV[0]

puts(":: Question: #{wolfram_question}")
puts("++ Getting Answer from WolframAlpha...")

wolfram_response = wolfram_client.query wolfram_question
wolfram_input = wolfram_response["Input"]
wolfram_result = wolfram_response.find { |pod| pod.title == "Result" }

puts(":: Answer:")
puts(wolfram_result.subpods[0].plaintext)

puts(":: Answer to synthesize:")
wolfram_answer = "for the question #{wolfram_question}, the interpretation is [ #{wolfram_input.subpods[0].plaintext} ], and the answer is #{wolfram_result.subpods[0].plaintext}"
puts wolfram_answer

puts("++ Getting synthesized audio file from Amazon Polly...")
polly_client = Aws::Polly::Client.new(region: "us-east-1")

polly_client.synthesize_speech({
  response_target: "#{Dir.pwd}/answer.pcm",
  output_format: "pcm",
  sample_rate: "16000",
  text: wolfram_answer,
  text_type: "text",
  voice_id: "Joanna",
})

exec("play -t raw -r 16k -e signed -b 16 -c 1 #{Dir.pwd}/answer.pcm")
